feat: auth-req caching

moolen opened the pull request:
**What this PR does / why we need it**:
This PR adds a way to configure the `proxy_cache_*` ([docs](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache_key)) directive for external authentication.

It adds these annotations, as proposed by @Stono in #2862 :
* `nginx.ingress.kubernetes.io/auth-cache-key`
* `nginx.ingress.kubernetes.io/auth-cache-duration`

The first one sets the `proxy_cache_key` (e.g. `$uri$cookie__auth_proxy`) and enables the feature. The latter sets the cache duration (defaults to `10m`)

**Note:**
The user-defined `cache_key` may contain sensitive information (e.g. `Authorization` header). That's why we want to store *only* a hash of that key, not the key itself on disk.


fixes #2862

aledbf commented:
@moolen please add e2e tests

aledbf commented:
/ok-to-test

aledbf commented:
/hold

moolen commented:
/retest

aledbf commented:
@moolen two comments:
- I think `auth-cache-key` should prepend the nginx server_name variable
- `auth-cache-duration` could be a list to allow different durations (not sure 10 minutes is ok for 401)

moolen commented:
Thanks for your feedback!

> I think auth-cache-key should prepend the nginx server_name variable

good catch, that makes sense!

> auth-cache-duration could be a list to allow different durations

You're right, it should be a list to allow configuration of multiple durations.

But while i think about that i'm not really sure if we really need this annotation at all. The user could set `proxy_cache_valid` using the `auth-snippet` annotation anyway. It would be semantically more precise to use a specific annotation, but adds more complexity to the codebase.

I'll implement it as a list for now. If you like using `auth-snippet` better i can remove it again (not a big deal). 

> not sure 10 minutes is ok for 401

It depends on the user's context i would say. For my use-case ($corp internal web-apps) i use `Authorization` header as cache_key and i want every auth-response (!= 5xx) to be cached.

I'd propose to lower the value to `5m`. What do you think?

moolen commented:
i did touch this; The last runs indicate this test is flaky; let's retest and maybe dig deeper.
/test pull-ingress-nginx-test

aledbf commented:
> I'll implement it as a list for now.

:+1: 

> I'd propose to lower the default value to 5m

:+1: 

Also, please squash the commits

aledbf commented:
/approve

ElvinEfendi commented:
> I think auth-cache-key should prepend the nginx server_name variable

should it include location identifier too? since one can configure different external authentication for different locations in the same server.

moolen commented:
again, that flaky test :/
/test pull-ingress-nginx-test

aledbf commented:
> again, that flaky test :/

Apologies for that :(
We are removing the static SSL feature in this release. With that, this will not be an issue anymore.

moolen commented:
@ElvinEfendi i added the [location identifier](https://github.com/kubernetes/ingress-nginx/pull/4278/files#diff-980db9e4b88704f12338bd074839f94eR917) and squashed the commits.

ElvinEfendi commented on docs/user-guide/nginx-configuration/configmap.md:
```diff
@@ -160,6 +160,8 @@ The following table shows a configuration option's name, type, and the default v
 |[global-auth-request-redirect](#global-auth-request-redirect)|string|""|
 |[global-auth-snippet](#global-auth-snippet)|string|""|
+|[global-auth-cache-key](#global-auth-cache-key)|string|""|
+|[global-auth-cache-duration](#global-auth-cache-duration)|string|"10m"|
```
why does this have a different default than https://github.com/kubernetes/ingress-nginx/pull/4278/files#diff-9bba411a7c28f1ef63c3a5339db109d5R385?

moolen replied:
that was wrong, thanks for catching that

ElvinEfendi commented on internal/ingress/controller/template/configmap.go:
```diff
@@ -226,6 +229,33 @@ func ReadConfig(src map[string]string) config.Configuration {
+					}
+				}
+			}
+		}
```
Is there any way to not duplicate this logic? @aledbf do we have means to share code between here and annotation? like a utils package or something

moolen replied:
I centralized the "parsing" logic for now in https://github.com/kubernetes/ingress-nginx/pull/4278/commits/d6b901419501b9bf711d650ac7dd0d9bee05b180. Let me know if you want it in a utils package. I digged into this topic - it seems there's a bunch of code that is copy/pasted with slight adaptions (e.g. returning vs. logging an error).

ElvinEfendi commented on rootfs/etc/nginx/template/nginx.tmpl:
```diff
@@ -894,6 +897,26 @@ stream {
+            proxy_cache auth_cache;
+
+            {{ if not $externalAuth.AuthCacheDuration }}
+            proxy_cache_valid 200 202 401 5m;
```
Can you make sure this is handled in the code, i.e by default `AuthCacheDuration` has this value. It's always better to avoid logic in template.

moolen replied:
I centralized it in https://github.com/kubernetes/ingress-nginx/pull/4278/commits/d6b901419501b9bf711d650ac7dd0d9bee05b180.

ElvinEfendi commented on rootfs/etc/nginx/template/nginx.tmpl:
```diff
@@ -894,6 +897,26 @@ stream {
+            set $cache_key '{{ $externalAuth.AuthCacheKey }}';
+
+            rewrite_by_lua_block {
+                local hash = ngx.sha1_bin(ngx.var.cache_key)
```
You can embed `$externalAuth.AuthCacheKey` directly here as a string instead of introducing a variable.

ElvinEfendi replied:
Oh I guess if you directly embed here then you'd have to implement concatenation of multiple variables yourself - this is good.

moolen replied:
If i embed `$externalAuth.AuthCacheKey` in a lua block i (as a user) can not use variables like `$uri` or `$cookie_name` via annotation or configMap.

If i use them before the lua block nginx replaces them with their values:
`$uri` becomes `/_external-auth-Lw` etc..

ElvinEfendi replied:
yup, that's what I figured after posting my comment (hence the second comment as an update)

not asking for any change here if that wasn't clear in my second comment

ElvinEfendi commented on rootfs/etc/nginx/template/nginx.tmpl:
```diff
@@ -894,6 +897,26 @@ stream {
+
+            rewrite_by_lua_block {
+                local hash = ngx.sha1_bin(ngx.var.cache_key)
+                ngx.var.cache_key = ngx.encode_base64(hash)
```
I wonder if we should introduce a new variable. You are overloading the variable with two different type of values.

moolen replied:
i'm mutating the value, yes. but the type stays from my understanding the same (string).

Adding an extra variable makes it unreadable IMO because we have to declare the variable beforehand:

```
set $tmp_cache_key '{{ ..things... }}';
set $real_cache_key ''; # <--- this

rewrite_by_lua {
  ngx.var.real_cache_key = nginx.encode_base64(ngx.sha1_bin(ngx.var.tmp_cache_key))
}

proxy_cache_key "$real_cache_key";

```

ElvinEfendi replied:
I don't feel strongly, feel free to not change this.

But subjectively I like the more verbose version you posted above.

moolen replied:
i changed to the slightly more verbose version.

ElvinEfendi commented:
Can you add a test to assert that the cache key includes location and server identifiers? For location the test would like like this

1. Sign in on i.e /foo
2. Kill the external auth server
3. Assert that /foo still works
3. Assert that /bar does not work

moolen commented:
Thanks for your feedback!

I added tests to verify the scope of the cache key (location/server) and addressed the other comments too.
I'm not sure about introducing a `utils` package. I centralized the logic for now. please let me know if that works for you! 

/test pull-ingress-nginx-test

moolen commented:
/test pull-ingress-nginx-test

moolen commented:
/test pull-ingress-nginx-e2e-1-12

ElvinEfendi commented on docs/user-guide/nginx-configuration/annotations.md:
```diff
@@ -378,6 +380,10 @@ Additionally it is possible to set:
 * `nginx.ingress.kubernetes.io/auth-request-redirect`:
   `<Request_Redirect_URL>`  to specify the X-Auth-Request-Redirect header value.
+* `nginx.ingress.kubernetes.io/auth-cache-key`:
+  `<Cache_Key>` this enables caching for auth requests. specify a lookup key for auth responses. e.g. `$remote_user$http_authorization`.
```
Should we make a note that this will be hashed and encoded?

moolen replied:
I think the implementation details are not important - the mechanics are tho!

I appended this:
> Each server and location has it's own keyspace. Hence a cached response is only valid on a per-server and per-location basis

ElvinEfendi commented:
@moolen this is looking great! Once you squash the commits we can merge this.

ElvinEfendi commented:
/test pull-ingress-nginx-e2e-1-12

ElvinEfendi commented:
/lgtm

thanks @moolen !

Stono commented:
Hey a quick one - thanks for all your effort in this everyone, finally able to rid it of my monkeypatch

vaskozl commented:
Isn't this dangerous in the case where the user specified cache key ends up being an empty variable? 

 E.g. if the cache key is set to `$cookie_foo` and the upstream auth provider decides to change the auth cookie name form `foo` to `bar`, then `$cookie_foo` would become empty and we start caching simply on `{ $server.Hostname }}{{ $authPath }}` which as I see it will let everyone through for the duration of the cache?
